Fixed the function to get the origin nodes of fused triton kernel.

shengfukevin opened the pull request:
Summary:
This DIFF is to fix the following issue:
In python source code for CompiledFxGraph,the FX graph segment for the Triton kernel is broken. For example, the following function
  def fn(a, b, c):
      x = torch.nn.functional.linear(a, b)
      x = x.sin()
      x = x.t() + c
      return x
Inductor compiled this FX graph into two nodes: the first one is mm, the second one is a triton kernel for sin + transpose + add. The FX graph segment for the triton kernel is like the following:
Graph fragment:
%add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %arg2_1), kwargs = {})
Basically only "add" node in the FX graph.
The root cause is function caffe2/torch/_inductor/utils.py:gather_origins does not detect the realized node correctly.
To fix this issue, the IRNode is checked if it is one of the following IRNode:
    ir.ComputedBuffer,
    ir.InputsKernel,
    ir.InputBuffer,
    ir.ReinterpretView,
    ir.TemplateBuffer,

If it is one of them, it is realized, otherwise, it is not.

Test Plan:
buck2 run mode/opt caffe2/test/inductor:provenance_tracing -- caffe2.test.inductor.test_provenance_tracing.TestProvenanceTracingArtifact.test_triton_kernel_to_post_grad_tracing_cuda

Rollback Plan:

Differential Revision: D77748371




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

shengfukevin commented:
@pytorchbot label "topic: not user facing"

mlazos reviewed (approved):
Had one nit comment on the diff to extract out the isinstance to a separate function, otherwise, LGTM.

mlazos commented:
Could you add a test as well?

shengfukevin commented:
> Could you add a test as well?

The origin fx nodes are used to define the name of triton kernel. There are some existing tests (test_aot_inductor, test_max_autotune) that check the name of the kernel within these tests. I have to update these name after fixing it with this PR. So I think this function has already been tested in these tests.

shengfukevin commented:
@mlazos, I made a few more changes after you reviewed it: 1. support the ops that has a container of tensors as the input (for example torch.cat). 2. Updated the unit tests that check the name of the kernel. Can you take another took? 

I checked the failed test 'test/test_reductions.py::TestReductionsCPU::test_sum_all_cpu_float64' is a known issue, not related to my PR.

Thanks

shengfukevin committed:
Fixed the function to get the origin nodes of fused triton kernel. (#157578)

Summary:

This DIFF is to fix the following issue:
In python source code for CompiledFxGraph,the FX graph segment for the Triton kernel is broken. For example, the following function
  def fn(a, b, c):
      x = torch.nn.functional.linear(a, b)
      x = x.sin()
      x = x.t() + c
      return x
Inductor compiled this FX graph into two nodes: the first one is mm, the second one is a triton kernel for sin + transpose + add. The FX graph segment for the triton kernel is like the following:
Graph fragment:
%add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%permute_1, %arg2_1), kwargs = {})
Basically only "add" node in the FX graph.
The root cause is function caffe2/torch/_inductor/utils.py:gather_origins does not detect the realized node correctly.
To fix this issue, the IRNode is checked if it is one of the following IRNode:
    ir.ComputedBuffer,
    ir.InputsKernel,
    ir.InputBuffer,
    ir.ReinterpretView,
    ir.TemplateBuffer,

If it is one of them, it is realized, otherwise, it is not.

Test Plan:
buck2 run mode/opt caffe2/test/inductor:provenance_tracing -- caffe2.test.inductor.test_provenance_tracing.TestProvenanceTracingArtifact.test_triton_kernel_to_post_grad_tracing_cuda

Rollback Plan:

Reviewed By: muchulee8, mlazos

Differential Revision: D77748371

shengfukevin commented on torch/_inductor/utils.py:
```diff
@@ -796,21 +796,36 @@ def dominated_nodes(
-    return OrderedSet(itertools.chain(*arg_origins, *kwarg_origins))
+        return isinstance(n, ir.IRNode) and not ir.IRNode.is_realized_node(n)
+
+    def get_unrealized_nodes(n: IRNode) -> OrderedSet[Any]:
```
Fixed. thanks
